highlighter: Add support for multiple symbols per identifier.

emilio opened the pull request:
This fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1510318 and
https://bugzilla.mozilla.org/show_bug.cgi?id=1430112.

There's a chance that this highlights too much for C++, but from my testing it
looks ok, and we can always filter symbols / add more heuristics if needed.

I removed some unused [data-kind] rules while I was at it.

staktrace commented on tools/src/format.rs:
```diff
@@ -140,15 +131,27 @@ pub fn format_code(jumps: &HashMap<String, Jump>, format: FormatAs,
                 let index = generated_json.len();
+                let syms = {
+                    let mut syms = String::new();
+                    for (i, sym) in d.iter().flat_map(|item| item.sym.iter()).enumerate() {
```
Can't you do `d.iter().flat_map(|item| item.sym.iter()).join(",")` ?

emilio replied:
Not without cloning all the symbols, but maybe that's ok.

staktrace commented:
So the dev instance is up. I'm actually seeing missing highlights for local variables, see around https://dev.searchfox.org/mozilla-central/source/gfx/layers/apz/src/APZCTreeManager.cpp#1044 for example - `clipRegion`, `aLayer`, `aParent` etc don't seem to highlight at all...

emilio commented on tools/src/format.rs:
```diff
@@ -102,15 +102,6 @@ pub fn format_code(jumps: &HashMap<String, Jump>, format: FormatAs,
-                let mut syms = d.iter().flat_map(|item| item.sym.iter());
-                let id = syms.find(|sym| sym.contains(displayed)).unwrap_or(&d[0].sym[0]);
-
                 let d = d.iter().filter(|item| { !item.no_crossref }).collect::<Vec<_>>();
```
Thanks, nice catch :)

staktrace commented:
Maybe we can further filter the results of `findReferences` and only select the `span`s where the textContent matches the textContent of the element the user is mousing over? That should then restrict it to tokens that appear the same to the user and hovering over an `EventStateManager` should only highlight other `EventStateManager`s.

emilio commented:
> Maybe we can further filter the results of findReferences and only select the spans where the textContent matches the textContent of the element the user is mousing over?

That works for types, but not for locals. For example, for the `aEvent` variable in that function you mentioned above, the symbols are: `V_0bd2a3ae_885764fe,V_c95394516f7eb61f_885764fe256`. So we can't just do that. We could try to strip out symbols for types (which in C++ start by `T_`) that don't match the `textContent` or something, but I'd rather merge this as-is.

emilio commented:
I fixed up the patch to build the symbol list before the `no_crossref` filtering, and also cleaned up a bit the surrounding code in a separate commit.

staktrace commented:
> For example, for the aEvent variable in that function you mentioned above, the symbols are: V_0bd2a3ae_885764fe,V_c95394516f7eb61f_885764fe256

I was thinking this filtering would use the user visible token, not the symbol. But anyway I'll experiment with that separately.

emilio commented:
Ah! Yeah, that makes sense to me.
